service/storage_proxy.cc

```diff
@@ -5960,6 +5959,97 @@ storage_proxy::query_partition_key_range(lw_shared_ptr<query::read_command> cmd,
     co_return coordinator_query_result(merger.get(), std::move(used_replicas));
 }
 
+future<result<storage_proxy::coordinator_query_result>>
+storage_proxy::query_partition_key_range_tablets(
+        schema_ptr schema,
+        locator::effective_replication_map_ptr erm,
+        lw_shared_ptr<query::read_command> cmd,
+        dht::partition_range_vector partition_ranges,
+        db::consistency_level cl,
+        storage_proxy::coordinator_query_options query_options) {
+    query_ranges_to_vnodes_generator ranges_to_vnodes(erm->make_splitter(), schema, std::move(partition_ranges), false);
+
+    auto trace_state = query_options.trace_state;
+    auto permit = query_options.permit;
+    auto timeout = query_options.timeout(*this);
+
+    std::vector<foreign_ptr<lw_shared_ptr<query::result>>> results;
+    auto p = shared_from_this();
+    auto& cf = _db.local().find_column_family(schema);
+    auto pcf = _db.local().get_config().cache_hit_rate_read_balancing() ? &cf : nullptr;
+    const auto& tm = erm->get_token_metadata();
+
+    if (_features.range_scan_data_variant) {
+        cmd->slice.options.set<query::partition_slice::option::range_scan_data_variant>();
+    }
+
+    auto& preferred_replicas = query_options.preferred_replicas;
+
+    const auto preferred_replicas_for_range = [&preferred_replicas, &tm] (const dht::partition_range& r) {
+        auto it = preferred_replicas.find(r.transform(std::mem_fn(&dht::ring_position::token)));
+        return it == preferred_replicas.end() ? inet_address_vector_replica_set{} : replica_ids_to_endpoints(tm, it->second);
+    };
+    const auto to_token_range = [] (const dht::partition_range& r) { return r.transform(std::mem_fn(&dht::ring_position::token)); };
+
+    dht::partition_range_vector ranges = ranges_to_vnodes(1);
+    dht::partition_range_vector::iterator i = ranges.begin();
+
+    if (i != ranges.end()) {
+        dht::partition_range& range = *i;
```

Comment:
I have been in this area recently, it seems the problem is that you're only consuming one range from query_ranges_to_vnodes_generator, even though partition_ranges, fed into the generator, may overlap with multiple tablet ranges. If you want to serialize execution across ranges, you need to only stop once all ranges are consumed.

Reply:
The client loops around pages, and I wanted to have only one loop.